docs/source/demos/punched_domain.geo.rst

```diff
@@ -0,0 +1,113 @@
+
+  Plane Surface(1) = {9, 10};
+  Plane Surface(2) = {10};
+
+Finally, we group together some edges and define ``Physical`` entities, which
+can be accessed in Firedrake. ::
+
+  Physical Curve("HorEdges", 11) = {1, 3};
+  Physical Curve("VerEdges", 12) = {2, 4};
+  Physical Curve("Circle", 13) = {8, 7, 6, 5};
+  Physical Surface("PunchedDom", 3) = {1};
+  Physical Surface("Disc", 4) = {2};
+
+For simplicity, we have gathered all this commands in the file
+`punched_domain.geo <punched_domain.geo>`__. To generate a mesh using this file,
+you can type the following command in the terminal::
+
+    gmsh -2 punched_domain.geo -format msh2
+
+To illustrate how to access all these features within Firedrake,
+we consider the following interface problem. Denoting by
+:math:`\Omega` the filled rectangle and by :math:`D` the disc,
+we seek a function :math:`u\in H^1_0(\Omega)` such that
+
+.. math::
+
+   -\nabla \cdot (\sigma \nabla  u) + u = 5 \quad \textrm{in } D
+
+where :math:`\sigma = 1` in :math:`\Omega \setminus D` and :math:`\sigma = 2`
+in :math:`D`. Since :math:`sigma` attains different values across :math:`\partial \Omega`,
+we need to prescribe the behavior of :math:`u` across the interface. This is
+implicitely done by imposing :math:`u\in H^1_0(\Omega)`: the function :math:`u` must be continuous
+across :math:`\partial \Omega`. This allows us to employ Lagrangian finite elements
+to approximate :math:`u`. However, something that we need to specify is the the jump
+of :math:`\sigma \nabla u \cdot \vec{n}` on :math:`\partial \Omega`. This terms arises
+natuarlly in the weak formulation of the problem under consideration. In this demo
+we simply set
+
+.. math::
+
+   [\sigma \nabla u \cdot \vec{n}] = 3 \quad \textrm{on}\ \partial D
+
+.. note::
+   In in practical applications, it is imporant to specify whether the normal vector
+   :math:`\vec{n}` points inward or outward. One should also specify the order in which
+   the jump operator is computed. This is irrelevant for the purpose of our demo and may
+   only lead to replacing the sign of boundary integral in the weak formulation.
+
+The resulting weak formulation reads as follows:
+
+.. math::
+
+   \int_D \sigma \nabla u \cdot \nabla v + uv \mathrm{d}\mathbf{x} - \int_{\partial D} 3v \mathrm{d}S = \int_{\Omega} 5v \mathrm{d}\mathbf{x} \quad \text{for every } v\in H^1_0(\Omega)\,.
+
+The following Firedrake code shows how to solve this variational problem
+using linear Lagrangian finite elements. ::
+
+   from firedrake import *
+   mesh = Mesh('punched_domain.msh')
+   V = FunctionSpace(mesh, "CG", 1)
+   u = TrialFunction(V)
+   v = TestFunction(V)
+   a = 2*dot(grad(v), grad(u))*dx((4,)) + dot(grad(v), grad(u))*dx((3,)) + v*u*dx
+   L = Constant(5.) * v * dx + Constant(3.)*v*ds((13,))
+   DirBC = DirichletBC(V, 0, [11, 12])
+   u = Function(V)
+   solve(a == L, u, bcs=DirBC, solver_parameters={'ksp_type': 'cg'})
```

Comment:
It is conventional for Firedrake demos to include a link to the generated python file at the bottom of the demo. See any of the other demos for an example of how to do this.

Reply:
I moved everything to `firedrake/demos/immersed_fem` and renamed the files ìmmersed_domain.geo` and `immersed_fem.py.rsi`

Reply:
But I see a potential issue: the file `immersed_fem.py.rst` contains both gmsh instructions and firedrake instructions. By looking at the automatically generated file `immersed_fem.py` in `firedrake/docs/source/demos`, I suspect that it would not run.

How should I fix this?